packages/firestore/src/local/shared_client_state.ts

```diff
@@ -409,15 +408,31 @@ export class WebStorageSharedClientState implements SharedClientState {
   }
 
   // TODO(multitab): Handle user changes.
-  start(initialUser: User, knownClients: ClientKey[]): void {
+  async start(initialUser: User): Promise<void> {
     assert(!this.started, 'WebStorageSharedClientState already started');
     assert(
       this.syncEngine !== null,
       'Start() called before subscribing to events'
     );
+
+    // We add the storage observer before we retrieve the list of existing
+    // clients to allow us to see LocalStorage notifications for clients
+    // that were added before getActiveClients() returns.
```

Comment:
When I started this PR, it also introduced the concept of `earlyEvents`, which I ended up leaving out (it only made it to the branch name). I have since added it back. These events start accumulating after the object is created, and get processed once the client is started. Capturing these events should address the race for both mutation batches and client states, but I would still like to keep the the call to `getActiveClients` as this interaction is much more obvious.

When I addressed this, I also ended up queueing all LocalStorage events on the AsyncQueue. This ensures that start() doesn't dispatch any events until after it gets returned and moves all local state access to the queue. It does, however, also mean that the test code now has to drain the queue. Instead of adding yet another argument to `testWebStorageSharedClientState` I moved the initialization to the test case itself. Let me know what you think and sorry about the churn :/